lib/hooks/trade.tsx

```diff
@@ -0,0 +1,238 @@
+  const { data: itemState } = useTradeItemState(item);
+
+  if (!itemState) {
+    return new Decimal(0);
+  }
+
+  const {
+    poolAssetBalance,
+    baseWeight,
+    poolBaseBalance,
+    assetWeight,
+    tradeablePoolAssetBalance,
+    swapFee,
+    traderAssetBalance,
+    traderBaseBalance,
+  } = itemState;
+
+  let maxAmountBase: Decimal;
+
+  if (item.action === "buy") {
+    maxAmountBase = calcInGivenOut(
+      poolBaseBalance,
+      baseWeight,
+      poolAssetBalance,
+      assetWeight,
+      tradeablePoolAssetBalance,
+      swapFee,
+    );
+  }
+  if (item.action === "sell") {
+    const assetBalance = traderAssetBalance.gt(tradeablePoolAssetBalance)
+      ? tradeablePoolAssetBalance
+      : traderAssetBalance;
+
+    maxAmountBase = calcOutGivenIn(
+      poolAssetBalance,
+      assetWeight,
+      poolBaseBalance,
+      baseWeight,
+      assetBalance,
+      swapFee,
+    );
+  }
+
+  maxAmountBase =
+    traderBaseBalance == null
+      ? new Decimal(0)
+      : maxAmountBase.gt(traderBaseBalance)
+      ? traderBaseBalance
+      : maxAmountBase;
+
+  return maxAmountBase;
+};
+
+export const useTradeMaxAssetAmount = (item: TradeItem): Decimal => {
+  const { data: itemState } = useTradeItemState(item);
+
+  const maxBaseAmount = useTradeMaxBaseAmount(item);
+
+  if (!itemState) {
+    return new Decimal(0);
+  }
+
+  const {
+    poolAssetBalance,
+    baseWeight,
+    poolBaseBalance,
+    assetWeight,
+    tradeablePoolAssetBalance,
+    swapFee,
+    traderBaseBalance,
+  } = itemState;
+
+  let maxAmountAsset: Decimal;
+
+  if (item.action === "buy") {
+    maxAmountAsset = calcOutGivenIn(
+      poolBaseBalance,
+      baseWeight,
+      poolAssetBalance,
+      assetWeight,
+      maxBaseAmount,
+      swapFee,
+    );
+  }
+  if (item.action === "sell") {
+    maxAmountAsset = calcInGivenOut(
+      poolAssetBalance,
+      assetWeight,
+      poolBaseBalance,
+      baseWeight,
+      maxBaseAmount,
+      swapFee,
+    );
+  }
+
+  maxAmountAsset =
+    traderBaseBalance == null
+      ? new Decimal(0)
+      : maxAmountAsset.gt(tradeablePoolAssetBalance)
+      ? tradeablePoolAssetBalance
+      : maxAmountAsset;
+
+  return maxAmountAsset;
+};
+
+export const useTradeTransaction = (
+  item: TradeItem,
+  amountAsset?: string,
+): SubmittableExtrinsic<"promise", ISubmittableResult> | undefined => {
+  amountAsset = !amountAsset ? "0" : amountAsset;
+  const amountAssetDecimal = new Decimal(amountAsset).mul(ZTG);
+  const [sdk] = useSdkv2();
+  const { data: itemState } = useTradeItemState(item);
+
+  if (itemState == null || sdk == null) {
+    return undefined;
+  }
+
+  const {
+    pool,
+    poolAssetBalance,
+    baseWeight,
+    poolBaseBalance,
+    assetWeight,
+    swapFee,
+    slippage,
+  } = itemState;
+
+  let transaction:
+    | SubmittableExtrinsic<"promise", ISubmittableResult>
+    | undefined;
+
+  if (isRpcSdk(sdk)) {
+    try {
+      if (item.action == "buy") {
```

Comment:
I'm not sure if this is quite right. I think the "Exact Amount" should be determined by the last input the user changed. Can be a separate ticket though